### PR TITLE
Add multiple list collector and lists validation for single section

### PR DESCRIPTION
**Pull request by petechd**
### PR Context
Non repeating sections are only allowed to contain single list collector and single list to be used by it. This adds new list collector and lists number validation for a section and new validation enforces these rules.

### Checklist

* [ ] eq-translations updated to support any new schema keys which need translation

### CODE REVIEW

**Comment by petechd** [comment]
[Iteration 1 - List collector and changes to section summary (5 SP)](https://trello.com/c/EuB99QjH/5321-iteration-1-list-collector-and-changes-to-section-summary-5-sp)

**Review by rmccar — APPROVED** [review]

File: schemas/questionnaire_v1.json [hunk 1]
```diff
@@ -241,11 +241,27 @@
...
+                      "items": {
+                        "$ref": "https://eq.ons.gov.uk/value_sources.json#/answer_value_source"
+                      }
+                    },
+                    "item_anchor_answer_id": {
+                      "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string"
+                    },
+                    "item_label": {
+                      "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string"
                     }
```
[the remark this answers is not in the document]
Comment (petechd) [reply schemas/questionnaire_v1.json]: Added descriptions now, thanks.

[the remark this answers is not in the document]
Comment (petechd) [reply schemas/questionnaire_v1.json]: Added, thanks.

File: schemas/questionnaire_v1.json [hunk 2]
```diff
@@ -256,6 +272,9 @@
               },
               "page_title": {
                 "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string"
+              },
+              "show_non_item_answers": {
+                "type": "boolean"
```
[the remark this answers is not in the document]
Comment (petechd) [reply schemas/questionnaire_v1.json:281]: Added descriptions.

**Comment by petechd** [comment]
> Haven't reviewed code yet but here are the list of validation we need of the top of my head. Ignore any that has been implemented.
> 
> * validate `related_answers` has valid answer ids which references answer_id that exist and are from the list collector block. Iteration 2.5 would extend that to allow from the repeating blocks.
> * `item_anchor_answer_id` is an answer id from the list collector block.
> * When displaying all answers for section, only 1 list collector per section.

I was double checking it and it seems that all of these are already there now, but will run through it once more.

File: schemas/questionnaire_v1.json [hunk 3]
```diff
@@ -241,11 +241,27 @@
                     "title": {
                       "$ref": "https://eq.ons.gov.uk/string_interpolation/definitions.json#/string_with_placeholders"
                     },
+                    "item_title": {
+                      "$ref": "https://eq.ons.gov.uk/string_interpolation/definitions.json#/string_with_placeholders"
+                    },
```
[the remark this answers is not in the document]
Comment (MebinAbraham) [reply schemas/questionnaire_v1.json]: I would remove this. We'll refactor it later to drop `block.summary`.

File: app/error_messages.py [hunk 4]
```diff
@@ -1,6 +1,17 @@
 DUMB_QUOTES_FOUND = "Found dumb quotes(s) in schema text"
 DUPLICATE_ID_FOUND = "Duplicate id found"
 FOR_LIST_NEVER_POPULATED = "for_list is not populated by any ListCollector blocks"
+MULTIPLE_LISTS = "Section has multiple lists"
+MULTIPLE_LIST_COLLECTORS = "Section has multiple ListCollector blocks"
```
Comment (MebinAbraham) [inline app/error_messages.py:5]: Same as above.

File: app/validators/sections/section_validator.py [hunk 5]
```diff
@@ -292,3 +296,117 @@ def _get_question_variant_fields_sets(variants):
...
+        blocks = self.questionnaire_schema.get_blocks(type="ListCollector")
+        for block in blocks:
+            if block["for_list"] not in lists:
+                lists.append(block["for_list"])
+        if len(lists) > 1:
+            return True
+
+    def validate_show_non_item_answers_when_items_key_exists(self, summary):
+        if summary.get("show_non_item_answers") and not summary.get("items"):
+            self.add_error(error_messages.ITEMS_NOT_PRESENT)
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:337]: This should be done in JSON. Don't need python validation for this.

File: app/validators/sections/section_validator.py [hunk 6]
```diff
@@ -292,3 +296,117 @@ def _get_question_variant_fields_sets(variants):
...
+
+    def validate_related_answers_when_show_non_item_answers_exists(self, summary):
+        if summary.get("show_non_item_answers"):
+            related_answers_present = False
+            if items := summary.get("items"):
+                for item in items:
+                    if item.get("related_answers"):
+                        related_answers_present = True
+                if not related_answers_present:
+                    self.add_error(error_messages.RELATED_ANSWERS_NOT_PRESENT)
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:347]: Relation answer should be optional?. May not want any related items, just a list of items with other non-item answers.

File: app/validators/sections/section_validator.py [hunk 7]
```diff
@@ -292,3 +296,117 @@ def _get_question_variant_fields_sets(variants):
...
+                list_collector_ids = []
+                blocks = self.questionnaire_schema.get_blocks(type="ListCollector")
+                for block in blocks:
+                    list_collector_ids.extend(
+                        iter(
+                            self.questionnaire_schema.get_list_collector_answer_ids(
+                                block["id"]
+                            )
+                        )
+                    )
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:370]: This is duplicated below. Create a map up front for list item answer id by list name.
You are checking all lists. We should check for the matching list to make this extensible.

File: app/error_messages.py [hunk 8]
```diff
@@ -1,6 +1,17 @@
...
+ITEMS_NOT_PRESENT = "Items field not present for summary"
+RELATED_ANSWERS_NOT_PRESENT = "Related_answers field not present for summary"
+RELATED_ANSWERS_NOT_IN_LIST_COLLECTOR = (
+    "Related_answers id not present in any list collector"
+)
+NO_LABEL_FOR_RELATED_ANSWER = "No label found for {answer_id}, only answers that support labels can be used as related answers"
```
Comment (MebinAbraham) [inline app/error_messages.py:11]:
```suggestion
NO_LABEL_FOR_RELATED_ANSWER = "No label found for answer '{answer_id}', only answers that support labels can be used as related answers"
```

File: app/error_messages.py [hunk 9]
```diff
@@ -1,6 +1,17 @@
...
+ITEM_ANCHOR_ANSWER_ID_NOT_IN_LIST_COLLECTOR = (
+    "Item anchor answer id not present in any list collector"
```
Comment (MebinAbraham) [inline app/error_messages.py:13]:
```suggestion
    "Item anchor answer id not present in any list collector for list name '{list_name}'"
```

File: schemas/questionnaire_v1.json [hunk 10]
```diff
@@ -246,6 +246,26 @@
...
+                      "type": "array",
+                      "minItems": 1,
+                      "items": {
+                        "$ref": "https://eq.ons.gov.uk/value_sources.json#/answer_value_source"
+                      },
+                      "description": "Ids of answers to be displayed on list summary"
+                    },
+                    "item_anchor_answer_id": {
+                      "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string",
+                      "description": "The main answer_id for the item title used as anchor in links on list summary"
```
Comment (MebinAbraham) [inline schemas/questionnaire_v1.json:260]:
```suggestion
                      "description": "The main answer_id for the item title used as anchor in summary links"
```

File: schemas/questionnaire_v1.json [hunk 11]
```diff
@@ -246,6 +246,26 @@
...
+                    },
+                    "item_label": {
+                      "$ref": "https://eq.ons.gov.uk/string_interpolation/definitions.json#/string_with_placeholders",
+                      "description": "The accompanying label for the item title on list summary"
```
Comment (MebinAbraham) [inline schemas/questionnaire_v1.json:264]: Worth adding `"examples": ["Name of company", "Full name"]`

File: app/validators/sections/section_validator.py [hunk 12]
```diff
@@ -292,3 +296,117 @@ def _get_question_variant_fields_sets(variants):
...
+            return True
+
+    def has_multiple_lists(self):
+        lists = []
+        blocks = self.questionnaire_schema.get_blocks(type="ListCollector")
+        for block in blocks:
+            if block["for_list"] not in lists:
+                lists.append(block["for_list"])
+        if len(lists) > 1:
+            return True
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:333]:
```suggestion
       return len({block.get("for_list") for block in blocks}) > 1
```

File: app/validators/sections/section_validator.py [hunk 13]
```diff
@@ -292,3 +296,117 @@ def _get_question_variant_fields_sets(variants):
...
+                return True
+
+    def has_multiple_list_collectors(self):
+        list_collector_count = 0
+        for group in self.schema_element.get("groups"):
+            for block in group.get("blocks"):
+                if block["type"] in ["ListCollector"]:
+                    list_collector_count += 1
+        if list_collector_count > 1:
+            return True
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:324]: Could be simplified with a set comprehension.

File: app/validators/sections/section_validator.py [hunk 14]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=too-many-public-methods
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:1]: Make the internal ones private?.

Comment (petechd) [reply app/validators/sections/section_validator.py]: Done, thanks!

File: app/validators/sections/section_validator.py (hunk 13 above)
Comment (petechd) [reply app/validators/sections/section_validator.py]: We need to do this for every group in schema element so I simplified it to list comprehension, same the one above, we only do this for a section not for the whole questionnaire schema.

File: schemas/questionnaire_v1.json [hunk 15]
```diff
@@ -246,6 +246,26 @@
...
+                    },
+                    "item_title": {
+                      "$ref": "https://eq.ons.gov.uk/string_interpolation/definitions.json#/string_with_placeholders",
+                      "description": "The literal item display name on list summary"
                     }
```
[the remark this answers is not in the document]
Comment (petechd) [reply schemas/questionnaire_v1.json:269]: Removed, thanks!

File: schemas/questionnaire_v1.json (hunk 11 above)
Comment (petechd) [reply schemas/questionnaire_v1.json:264]: Added examples, thanks!

File: schemas/questionnaire_v1.json [hunk 16]
```diff
@@ -246,6 +246,23 @@
...
+                      "items": {
+                        "$ref": "https://eq.ons.gov.uk/value_sources.json#/answer_value_source"
+                      },
+                      "description": "Ids of answers to be displayed on list summary"
+                    },
+                    "item_anchor_answer_id": {
+                      "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string",
+                      "description": "The main answer_id for the item title used as anchor in summary links"
+                    },
+                    "item_label": {
```
Comment (MebinAbraham) [inline schemas/questionnaire_v1.json:262]: This new key will need to be added to `eq-translations`.

File: app/validators/sections/section_validator.py [hunk 17]
```diff
@@ -292,3 +295,121 @@ def _get_question_variant_fields_sets(variants):
...
+            self.has_list_summary_with_non_item_answers()
+            and self.has_multiple_list_collectors()
+        ):
+            self.add_error(error_messages.MULTIPLE_LIST_COLLECTORS)
+
+    def has_list_summary_with_non_item_answers(self):
+        if summary := self.schema_element.get("summary"):
+            show_non_item_answers = summary.get("show_non_item_answers")
+            if summary.get("items") and show_non_item_answers is True:
+                return True
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:314]:
```suggestion
            return summary.get("items") and show_non_item_answers
```

File: app/validators/sections/section_validator.py [hunk 18]
```diff
@@ -292,3 +295,121 @@ def _get_question_variant_fields_sets(variants):
...
+    def has_multiple_lists(self):
+        lists = []
+        for group in self.schema_element.get("groups"):
+            lists.extend(
+                block.get("for_list")
+                for block in group.get("blocks")
+                if block["type"] in ["ListCollector"]
+            )
+
+        return len(lists) > 1
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:336]: Both of these tell you the same thing, you only need 1 of them?. The `has_multiple_lists` check can be dropped as if we verify you only have 1 list collector, then that implies you also only have 1 list.

File: app/validators/sections/section_validator.py (hunk 6 above)
Comment (berroar) [reply app/validators/sections/section_validator.py]: Removed this check related answers are optional 👍

File: app/error_messages.py [hunk 19]
```diff
@@ -1,6 +1,17 @@
 DUMB_QUOTES_FOUND = "Found dumb quotes(s) in schema text"
 DUPLICATE_ID_FOUND = "Duplicate id found"
 FOR_LIST_NEVER_POPULATED = "for_list is not populated by any ListCollector blocks"
+MULTIPLE_LISTS = "Section has multiple lists"
```
[the remark this answers is not in the document]
Comment (berroar) [reply app/error_messages.py]: This has been removed now 👍

File: app/error_messages.py (hunk 4 above)
Comment (berroar) [reply app/error_messages.py]: Updated to be more meaningful 👍

File: schemas/questionnaire_v1.json [hunk 20]
```diff
@@ -246,6 +246,26 @@
...
                     "empty_list_text": {
                       "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string"
+                    },
+                    "related_answers": {
+                      "type": "array",
+                      "minItems": 1,
+                      "items": {
+                        "$ref": "https://eq.ons.gov.uk/value_sources.json#/answer_value_source"
+                      },
+                      "description": "Ids of answers to be displayed on list summary"
```
[the remark this answers is not in the document]
Comment (berroar) [reply schemas/questionnaire_v1.json]: Updated this 👍

File: app/validators/sections/section_validator.py (hunk 7 above)
Comment (berroar) [reply app/validators/sections/section_validator.py:340]: Updated this method 👍

File: app/validators/sections/section_validator.py (hunk 5 above)
Comment (MebinAbraham) [reply app/validators/sections/section_validator.py]: I don't think it is quite right .`dependencies` ensure that `items` must exist when `show_non_item_answers`, but it should be `items` must exist when `show_non_item_answers` is true. You need to check the value, so you need to do it via `if` `then`.

File: app/validators/sections/section_validator.py (hunk 7 above)
Comment (MebinAbraham) [reply app/validators/sections/section_validator.py:340]: It is still duplicated in the both functions. I would create the map and assigned in on line 324 so you only build the map if you need it.

In this method, the list name isn't checked either.

File: schemas/questionnaire_v1.json [hunk 21]
```diff
@@ -246,6 +246,23 @@
                     },
                     "empty_list_text": {
                       "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string"
+                    },
+                    "related_answers": {
+                      "type": "array",
+                      "minItems": 1,
```
Comment (MebinAbraham) [inline schemas/questionnaire_v1.json:252]:
```suggestion
                      "minItems": 1,
                      "uniqueItems": true,
```

File: schemas/questionnaire_v1.json [hunk 22]
```diff
@@ -246,6 +246,23 @@
...
+                      "items": {
+                        "$ref": "https://eq.ons.gov.uk/value_sources.json#/answer_value_source"
+                      },
+                      "description": "Value sources of answers to be displayed on the list summary, excluding the answer used for the item title"
+                    },
+                    "item_anchor_answer_id": {
+                      "$ref": "https://eq.ons.gov.uk/common_definitions.json#/non_empty_string",
```
Comment (MebinAbraham) [inline schemas/questionnaire_v1.json:259]:
```suggestion
                      "$ref": "https://eq.ons.gov.uk/common_definitions.json#/identifier"
```

File: schemas/questionnaire_v1.json (hunk 16 above)
Comment (berroar) [reply schemas/questionnaire_v1.json:263]: Added a PR 👍

File: app/validators/sections/section_validator.py (hunk 5 above)
Comment (berroar) [reply app/validators/sections/section_validator.py]: Updated the logic for this 👍

File: app/validators/sections/section_validator.py (hunk 7 above)
Comment (berroar) [reply app/validators/sections/section_validator.py:349]: I moved the map to line 324, but not sure this is what you intended? (checking the list name on line 364)

**Review by rmccar — APPROVED** [review]

File: app/validators/sections/section_validator.py [hunk 23]
```diff
@@ -292,3 +294,97 @@ def _get_question_variant_fields_sets(variants):
...
+                    answer = self.questionnaire_schema.get_answer(
+                        related_answer["identifier"]
+                    )
+                    if not answer.get("label"):
+                        self.add_error(
+                            error_messages.NO_LABEL_FOR_RELATED_ANSWER.format(
+                                answer_id=related_answer["identifier"],
+                            ),
+                            id=related_answer["identifier"],
+                        )
```
Comment (MebinAbraham) [inline app/validators/sections/section_validator.py:390]: The duplication and overall validation can be simplified with something like:
```suggestion
    def validate_section_summary(self):
        summary_items = self.schema_element.get("summary", {}).get("items", [])
        if not summary_items:
            return None

        blocks = self.questionnaire_schema.get_blocks(type="ListCollector")
        list_collector_answer_ids_by_list = defaultdict(list)
        for block in blocks:
            list_collector_answer_ids_by_list["for_list"].append(
                self.questionnaire_schema.get_list_collector_answer_ids(block["id"])
            )

        for item in summary_items:
            list_collector_answer_ids_for_list = list_collector_answer_ids_by_list[
                item["for_list"]
            ]

            if item_anchor_answer_id := item.get("item_anchor_answer_id"):
                self._validate_item_anchor_answer_id_belongs_to_list_collector(
                    item_anchor_answer_id, list_collector_answer_ids_for_list
                )

            for answer_source in item.get("related_answers", []):
                self._validate_related_answer_belong_to_list_collector(
                    answer_source, list_collector_answer_ids_for_list
                )
                self._validate_related_answer_has_label(answer_source)

    def _validate_related_answer_belong_to_list_collector(
        self, answer_source, list_collector_answer_ids
    ):
        if answer_source["identifier"] not in list_collector_answer_ids:
            self.add_error(
                error_messages.RELATED_ANSWERS_NOT_IN_LIST_COLLECTOR,
                id=answer_source["identifier"],
            )

    def _validate_item_anchor_answer_id_belongs_to_list_collector(
        self, anchor_answer_id, list_collector_answer_ids
    ):
        if anchor_answer_id not in list_collector_answer_ids:
            self.add_error(
                error_messages.ITEM_ANCHOR_ANSWER_ID_NOT_IN_LIST_COLLECTOR,
                id=anchor_answer_id,
            )

    def _validate_related_answer_has_label(self, answer_source):
        answer = self.questionnaire_schema.get_answer(answer_source["identifier"])
        if not answer.get("label"):
            self.add_error(
                error_messages.NO_LABEL_FOR_RELATED_ANSWER.format(
                    answer_id=answer_source["identifier"],
                ),
                id=answer_source["identifier"],
            )
```

Comment (MebinAbraham) [reply app/validators/sections/section_validator.py:390]: I suggest renaming `validate_section_summary` to `validate_section_summary_items`.

File: app/validators/sections/section_validator.py (hunk 7 above)
Comment (MebinAbraham) [reply app/validators/sections/section_validator.py:349]: Not quite, but have left a comment on the suggestion.

File: app/validators/sections/section_validator.py (hunk 23 above)
Comment (berroar) [reply app/validators/sections/section_validator.py]: Added now 👍

**Review by MebinAbraham — APPROVED** [review]

**Pull request merged**